lewis/adapters/modbus.py

```diff
@@ -0,0 +1,569 @@
+# -*- coding: utf-8 -*-
+# *********************************************************************
+# lewis - a library for creating hardware device simulators
+# Copyright (C) 2016 European Spallation Source ERIC
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+# References Used:
+#   http://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf
+#   http://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf
+#   https://github.com/sourceperl/pyModbusTCP
+#   https://github.com/bashwork/pymodbus
+# *********************************************************************
+
+
+from __future__ import division
+
+import socket
+import asyncore
+import struct
+
+from copy import deepcopy
+from math import ceil
+from argparse import ArgumentParser
+
+from lewis.adapters import Adapter
+
+
+class ModbusDataBank(object):
+    def __init__(self, config):
+        """
+        Preliminary DataBank implementation for Modbus.
+
+        This is a very generic implementation of a databank for Modbus. It's meant to set the
+        groundwork for future implementations. The create_basic method should be used for
+        construction, rather than this initializer. The signature of this __init__ method is
+        subject to change.
+
+        .. sourcecode:: Python
+
+            di = ModbusDataBank.create_basic(False, 0x1000, 0x1FFF)
+
+        :param config: dict containing configuration
+        """
+        self._data = config['data']
+        self._start_addr = config['start_addr']
+
+    @classmethod
+    def create_basic(cls, default_value=0, start_addr=0x0000, last_addr=0xFFFF):
```

Comment:
>This sounds like an argument for sub-classing to me.

Only if they need to do different things for different DataBanks. I don't think that would be the case. But `_data` would have to become a `dict` instead of a `list` and so `get` and `set` would have to change accordingly. Right now, this just isn't necessary.

But, yes, exactly the same can be achieved with inheritance... it's just a question of whether you have many subclasses or many `create_type` methods. Subclassing does have the (huge) advantage that you have the option of changing behaviour completely if you need to... the `create` methods just felt like a more tidy alternative.